### PR TITLE
61 - create component to select multiple users

### PR DESCRIPTION
**Pull request by Its-OP**
Component for multiple user selection was added.

**Part 1**
- The component is responsive. Recommended width-span is 400-1000px without roles or 600-1300px with roles.
- The component is customizable. It takes an object of type `IBoard` on initialization (could be extended if needed) and works around it.
- Email input gets validated before the data is sent to the service.
- Usernames in the list are clickable. They are supposed to redirect to the user's profile.
- Loader component is integrated in the select-user component. Alas, our loader component is not finished by now, so it just shows a text.
- Notification when there are no users in the board was implemented.
- Pagination was not implemented, as I doubt that there would be possible to add more than a hundred users to whatever project. However, the list limits amount of concurrently displayed users to 5, turning scroller on when there are more.
- A service was added for this component. At the moment it works by communicating with local storage. When our project controllers will be done, the service's implementation should be rewritten, however, the interace should stay intact.

**Part 2**
At the beginning, I thought that this component would interact with users primarily. As it turned out later, it is more efficient to interact with boards. While working on Back-End, I've created some infrastructure for `User` model, which was not required later. Despite it, I decided to leave it, as it may be useful in the future.

**Part 3**
Identity- back-end infrastructure was separated form BLL and now is a standalone class library.

### CODE REVIEW

File: backend/Tasque.WebAPI/Program.cs [hunk 1]
```diff
@@ -24,13 +25,24 @@
...
+AppConfigurationExtension.RegisterServices(builder.Services, builder.Configuration);
+
+builder.Services.RegisterIdentity();
 
 var app = builder.Build();
 
+using (var scope = app.Services.CreateScope())
+{
+    var dbContext = scope.ServiceProvider.GetRequiredService<DataContext>();
+    AppContext.SetSwitch("Npgsql.EnableLegacyTimestampBehavior", true);
```
Comment (Anatolii-Ivanov) [inline backend/Tasque.WebAPI/Program.cs]: Do we really need this?

File: frontend/src/shared/components/select-users/select-users.component.ts [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+      }
+    };
+    const board = getBoard();
+    const key = this.service.createKey(board);
+
+    if (localStorage.getItem(key)) {
+      this.board = JSON.parse(localStorage.getItem(key) as string);
+    } else {
+      this.board = board;
+    }
```
Comment (Anatolii-Ivanov) [inline frontend/src/shared/components/select-users/select-users.component.ts:74]: Can you make it more like a component? So, it won't load data from localstorage or send tons of http requests

File: backend/Tasque.WebAPI/Program.cs (hunk 1 above)
Comment (Its-OP) [reply backend/Tasque.WebAPI/Program.cs]: We do not, it has accidentally leaked in the PR

File: frontend/src/shared/components/select-users/select-users.component.ts (hunk 2 above)
Comment (Its-OP) [reply frontend/src/shared/components/select-users/select-users.component.ts:74]: These strings are here only for testing purpose. The board should always be provided to component, so this part would be deleted as soon as this component will be included anywhere.

Comment (Its-OP) [reply frontend/src/shared/components/select-users/select-users.component.ts:74]: P.S. `Observable<any>` will be changed to `Observable<TModel>` as well

File: backend/Identity/Registrations.cs [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+namespace Tasque.Core.Identity
+{
+    public static class Registrations
+    {
+        public static void RegisterIdentity(this IServiceCollection services)
+        {
+            services.AddScoped<IUserService, UserService>();
+        }
+    }
+}
```
Comment (FeloniukM) [inline backend/Identity/Registrations.cs]: white space

File: frontend/src/app/app.component.html [hunk 4]
```diff
@@ -1 +1 @@
-<router-outlet></router-outlet>
+<router-outlet></router-outlet>
```
Comment (FeloniukM) [inline frontend/src/app/app.component.html]: same

File: frontend/src/shared/components/select-users/select-users.component.sass [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+									border-width: 1px
+								
+								::ng-deep .mat-form-field-infix
+									padding-bottom: 0.3em
+
+								::ng-deep .mat-form-field-wrapper
+									padding-bottom: 0px
+
+							.mat-button-action
+								margin: 10px 15px
```
Comment (FeloniukM) [inline frontend/src/shared/components/select-users/select-users.component.sass]: same

File: backend/Tasque.WebAPI/appsettings.json [hunk 6]
```diff
@@ -12,9 +12,6 @@
     "Key": "BSA-2022-Tasque.Core"
   },
   "AllowedHosts": "*",
-  "ConnectionStrings": {
```
Comment (ismv) [inline backend/Tasque.WebAPI/appsettings.json]: are you sure you want to delete this?

Comment (Its-OP) [reply backend/Tasque.WebAPI/appsettings.json]: Yep. We hold connection strings in the user's secrets. This is just a template.

File: backend/Identity/Registrations.cs (hunk 3 above)
Comment (Its-OP) [reply backend/Identity/Registrations.cs]: Resolved all

File: backend/Tasque.Core.Identity/Services/Abstraction/IUserService.cs [hunk 7]
```diff
@@ -0,0 +1,10 @@
+﻿using Tasque.Core.BLL.Services;
+using Tasque.Core.Common.Entities;
+
+namespace Tasque.Core.Identity.Services.Abstraction
+{
+    public interface IUserService : IEntityCrudService<User>
+    {
+        User? GetByEmail(string email);
+    }
+}
```
Comment (ihor-il) [inline backend/Tasque.Core.Identity/Services/Abstraction/IUserService.cs]: eof

File: frontend/src/shared/components/select-users/select-users.component.sass (hunk 5 above)
Comment (ihor-il) [inline frontend/src/shared/components/select-users/select-users.component.sass:140]: Is such indentation intended?

**Review by ihor-il — CHANGES_REQUESTED** [review]
Remove .vs folder please

File: backend/Tasque.Core.Identity/Services/Abstraction/IUserService.cs (hunk 7 above)
Comment (Its-OP) [reply backend/Tasque.Core.Identity/Services/Abstraction/IUserService.cs]: Resolved

File: backend/Tasque.Core.WebAPI/Program.cs [hunk 8]
```diff
@@ -24,11 +25,13 @@
 builder.Services.AddCors();
 
 builder.Services.AddDbContext<DataContext>(
-    o => o.UseNpgsql(builder.Configuration.GetConnectionString("TasqueDb"),
+    o => o.UseNpgsql(builder.Configuration["ConnectionStrings:TasqueDb"], 
         b => b.MigrationsAssembly(typeof(DataContext).Assembly.FullName))
         .EnableDetailedErrors());
 
-builder.WebHost.UseUrls("http://*:5000");
```
Comment (ihor-il) [inline backend/Tasque.Core.WebAPI/Program.cs:32]: @Anatolii-Ivanov says that we need this line for deployment

File: backend/Tasque.Core.WebAPI/appsettings.json [hunk 9]
```diff
@@ -12,9 +12,6 @@
     "Key": "BSA-2022-Tasque.Core"
   },
   "AllowedHosts": "*",
-  "ConnectionStrings": {
```
Comment (ihor-il) [inline backend/Tasque.Core.WebAPI/appsettings.json]: Why is this line removed?

**Review by ihor-il — CHANGES_REQUESTED** [review]

File: frontend/src/shared/components/select-users/select-users.component.sass (hunk 5 above)
Comment (Its-OP) [reply frontend/src/shared/components/select-users/select-users.component.sass:140]: Yep. It is done to prevent possible appliances of styles to elements of other components.

File: backend/Tasque.Core.WebAPI/Program.cs (hunk 8 above)
Comment (Its-OP) [reply backend/Tasque.Core.WebAPI/Program.cs:32]: Resolved. I've deleted it by mistake.

File: backend/Tasque.Core.WebAPI/appsettings.json (hunk 9 above)
Comment (Its-OP) [reply backend/Tasque.Core.WebAPI/appsettings.json]: We do not really need it, as we store this data in user's secrets. But I can return it if you want.

**Review by Anatolii-Ivanov — DISMISSED** [review]

**Review by ismv — DISMISSED** [review]

**Review by ihor-il — DISMISSED** [review]

**Review by Anatolii-Ivanov — APPROVED** [review]

**Review by ismv — APPROVED** [review]
